internal/http/services/owncloud/ocgraph/drive_permissions.go

```diff
@@ -562,38 +567,90 @@ func (s *svc) writePermissions(ctx context.Context, w http.ResponseWriter, actio
 	}
 }
 
-func (s *svc) getLinkUpdate(ctx context.Context, permission *libregraph.Permission, resourceType providerpb.ResourceType) (*linkv1beta1.UpdatePublicShareRequest_Update, error) {
+func (s *svc) getLinkUpdates(ctx context.Context, link *linkv1beta1.PublicShare, permission *libregraph.Permission, resourceType providerpb.ResourceType) ([]*linkv1beta1.UpdatePublicShareRequest_Update, error) {
+	updates := []*linkv1beta1.UpdatePublicShareRequest_Update{}
+
+	defaultExpirationDefined := time.Second * time.Duration(s.c.PubRWLinkDefaultExpiration)
+	maxExpirationDefined := time.Second * time.Duration(s.c.PubRWLinkMaxExpiration)
+	endOfDay := time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 23, 59, 59, 0, time.Now().Location())
+	defaultExpiration := endOfDay.Add(defaultExpirationDefined)
+	maxExpiration := endOfDay.Add(maxExpirationDefined)
+
+	isExpirationEnforced := maxExpirationDefined > 0
+	isEditorLink := false
+	if permission.Link != nil && permission.Link.Type != nil {
+		isEditorLink = permission.Link.GetType() == libregraph.EDIT
+	} else if link.Permissions != nil {
+		isEditorLink = conversions.RoleFromResourcePermissions(link.Permissions.Permissions).Name == conversions.RoleEditor
+	}
+
 	if permission.ExpirationDateTime.IsSet() {
-		return &linkv1beta1.UpdatePublicShareRequest_Update{
+		finalExpiration := permission.ExpirationDateTime
+		if isEditorLink && isExpirationEnforced {
+			if permission.ExpirationDateTime.Get() == nil {
+				finalExpiration.Set(&defaultExpiration)
+			} else if permission.ExpirationDateTime.Get().After(maxExpiration) {
+				finalExpiration.Set(&maxExpiration)
+			}
+		}
+		updates = append(updates, &linkv1beta1.UpdatePublicShareRequest_Update{
 			Type: linkv1beta1.UpdatePublicShareRequest_Update_TYPE_EXPIRATION,
 			Grant: &linkv1beta1.Grant{
-				Expiration: nullableTimeToCs3Timestamp(permission.ExpirationDateTime),
+				Expiration: nullableTimeToCs3Timestamp(finalExpiration),
 			},
-		}, nil
-	} else if permission.Link != nil && permission.Link.LibreGraphDisplayName != nil {
-		if *permission.Link.LibreGraphDisplayName == "" {
+		})
+	}
+	if permission.Link != nil && permission.Link.LibreGraphDisplayName != nil {
```

Comment:
You don't have `else if` any longer because we cannot enter in more than one if, is this correct and certain?